Stale opencode wrapper at /usr/local/bin/opencode silently corrupts auth.json on every spawn (escapes upgrade.sh)

## TL;DR

`runtimes/opencode.sh::_install_opencode_wrapper` installs a bash wrapper at `/usr/local/bin/opencode` that pre-syncs Anthropic credentials before exec'ing the real binary. On any install where:

1. wp-coding-agents was originally configured with `RUNTIME=opencode`, then later switched to `RUNTIME=claude-code` (or any non-opencode runtime), AND
2. The wrapper is still being used by kimaki (kimaki resolves `opencode` via PATH and gets the wrapper),

the wrapper **rots** indefinitely — `upgrade.sh` will never replace or remove it, even though the wrapper format has evolved significantly since installation. On extrachill.com VPS the deployed wrapper is a 21-line March 22 version that reads from a `.credentials.json` file last-modified **six weeks ago** (also stale) and overwrites the kimaki-managed `auth.json` with that stale data **on every kimaki spawn of opencode-serve**.

User-visible symptom: every `systemctl restart kimaki` produces an `invalid_grant` OAuth crisis on the next prompt, requiring the user to `/login` again. Confirmed reproducible across multiple restarts.

## Repro on extrachill.com VPS (live)

```
$ ls -la /usr/local/bin/opencode
-rwxr-xr-x 1 root root 699 Mar 22 15:43 /usr/local/bin/opencode
$ wc -l /usr/local/bin/opencode
21 /usr/local/bin/opencode
$ stat /home/opencode/.claude/.credentials.json
Modify: 2026-03-22 15:34:18 (six weeks old)
```

Wrapper content:

```bash
#!/usr/bin/env bash
set -euo pipefail
AUTH_SRC="${HOME}/.claude/.credentials.json"
AUTH_DST="${HOME}/.local/share/opencode/auth.json"
if [[ -f "$AUTH_SRC" ]]; then
  if command -v jq >/dev/null 2>&1; then
    mkdir -p "$(dirname "$AUTH_DST")"
    jq "{anthropic: {type: \"oauth\", refresh: ..., access: ..., expires: ...}}" "$AUTH_SRC" > "${AUTH_DST}.tmp" 2>/dev/null && mv "${AUTH_DST}.tmp" "$AUTH_DST"
  fi
fi
exec /usr/lib/node_modules/opencode-ai/bin/opencode "$@"
```

Compare to the **current** `runtimes/opencode.sh` wrapper (~100 lines), which reads from kimaki's `anthropic-oauth-accounts.json` and conditionally updates only when kimaki has fresher creds. The deployed wrapper has none of that — it unconditionally clobbers `auth.json` with whatever (stale) state `.credentials.json` happens to have.

Cross-referencing with kimaki.log:

```
15:04:26  [OPENCODE] Resolved opencode binary: /usr/local/bin/opencode  ← wrapper
15:04:28  [OPENCODE] Spawned opencode serve --port 46785 (pid: 3910291)
          ↑ wrapper ran here, overwrote auth.json with stale March creds
15:14:17  [ERROR] HTTP 400 invalid_grant ← first refresh attempt fails
15:14:23  user runs /login
15:14:38  kimaki writes fresh creds to auth.json
          ↑ confirmed by mtime on auth.json
```

The wrapper hot-path is gated by kimaki's binary resolution, not by `which opencode`. Kimaki uses Node's `child_process.spawn('opencode', ...)` which goes through PATH — first match wins, and `/usr/local/bin/opencode` is ahead of `/usr/lib/node_modules/opencode-ai/bin/opencode` in the kimaki.service `Environment=PATH=` ordering. So **every kimaki restart that re-resolves the binary triggers the wrapper**.

## Two cascading upstream bugs

### Bug 1 — Phase 7 gate is on `RUNTIME`, not `CHAT_BRIDGE`

`upgrade.sh:629`:

```bash
if [ "$RUNTIME" != "opencode" ]; then
  log "Phase 7: Skipping (runtime is $RUNTIME, not opencode)"
  return 0
fi
```

But the wrapper exists for **kimaki**, not for **opencode-runtime**. Kimaki spawns opencode-serve regardless of which editor runtime is configured. `_install_opencode_wrapper` at `runtimes/opencode.sh:21-25` already gates on `CHAT_BRIDGE != kimaki` correctly:

```bash
if [ "$CHAT_BRIDGE" != "kimaki" ]; then
  return
fi
```

But Phase 7 in `upgrade.sh` short-circuits before that runtime file is even sourced, so the inner gate never gets a chance to evaluate. The result: claude-code-runtime + kimaki-bridge installs are stuck with whatever wrapper got installed under their previous runtime configuration, with no upgrade path.

### Bug 2 — `_install_opencode_wrapper` skips updates of existing wrappers

`runtimes/opencode.sh:36-40`:

```bash
# Don't wrap if opencode is already a wrapper script (re-runs)
if head -1 "$OPENCODE_BIN" 2>/dev/null | grep -q "bash"; then
  log "OpenCode wrapper already installed — skipping"
  return
fi
```

"Is the first line a bash shebang? Skip." That's it. No version check, no content diff against the current `WRAPPER_CONTENT`, no replacement. Once a wrapper is installed, it lives forever.

So even if Bug 1 is fixed and Phase 7 *does* run for kimaki+claude-code installs, **the existing stale wrapper still wouldn't be replaced** — the installer would log "already installed" and bail.

## Proposed fixes

### Fix Bug 1 — repair Phase 7 gating

Either:
- Drop the `RUNTIME` gate in Phase 7 entirely; let `runtimes/opencode.sh` decide internally based on `CHAT_BRIDGE`. That's cleanest — Phase 7 always sources the file and lets the installer self-skip when not applicable.
- Or change the gate to `if [ "$RUNTIME" != "opencode" ] && [ "$CHAT_BRIDGE" != "kimaki" ]` — keeps the short-circuit but covers the kimaki case.

The first form is preferable since the inner installer is already correctly gated.

### Fix Bug 2 — replace stale wrappers

Compare the installed wrapper's content to the canonical `WRAPPER_CONTENT` string. If they differ, replace. Conservatively: write a sentinel comment line into the wrapper (e.g. `# wp-coding-agents wrapper v2`) and replace if the version doesn't match. Idempotent re-runs continue to no-op when content matches.

```bash
# Pseudocode
EXPECTED_SENTINEL="# wp-coding-agents-wrapper-v2"
if grep -qF "$EXPECTED_SENTINEL" "$OPENCODE_BIN" 2>/dev/null; then
  log "OpenCode wrapper already at current version — skipping"
  return
fi

if head -1 "$OPENCODE_BIN" 2>/dev/null | grep -q "bash"; then
  log "OpenCode wrapper outdated — backing up and replacing"
  cp "$OPENCODE_BIN" "${OPENCODE_BIN}.bak.$(date +%Y%m%d)"
  # ...fall through to write new wrapper...
fi
```

## Interim workaround (applied to extrachill.com VPS today)

```bash
cp /usr/local/bin/opencode /usr/local/bin/opencode.bak.20260503
rm /usr/local/bin/opencode
```

After removal, `command -v opencode` resolves via PATH to `/usr/bin/opencode` (and from there, the real binary at `/usr/lib/node_modules/opencode-ai/bin/opencode`). Kimaki's auth plugin becomes the sole writer to `auth.json`, no more stale-data clobbering.

This works as an interim because:
- `CHAT_BRIDGE=kimaki` already manages OAuth refresh end-to-end via its own anthropic-auth-plugin
- The wrapper's *intent* (cred sync from `.credentials.json` to `auth.json`) is redundant when kimaki is the auth manager
- The wrapper's *side effect* (overwriting `auth.json` with stale data) is purely harmful

But this is a workaround, not a fix — the next `upgrade.sh` could potentially reinstall a wrapper if the cascade ever reaches `_install_opencode_wrapper` again.

## Acceptance criteria

- [ ] Phase 7 gating fixed so it doesn't skip when `CHAT_BRIDGE=kimaki` regardless of `RUNTIME`
- [ ] `_install_opencode_wrapper` detects + replaces stale wrappers (versioned sentinel or content diff)
- [ ] Existing installs with stale wrappers get them replaced on next `upgrade.sh` run
- [ ] Smoke test: install with `RUNTIME=opencode + CHAT_BRIDGE=kimaki`, switch to `RUNTIME=claude-code`, run upgrade, confirm wrapper at `/usr/local/bin/opencode` matches the current `WRAPPER_CONTENT` byte-for-byte

## Severity

**High.** Affects any install that switched runtimes after initial setup, with the symptom presenting as "OAuth fails on every kimaki restart." The user-facing experience is constant `/login` cycling. Diagnosis is non-obvious because:
- The wrapper is invisible (kimaki's process tree shows the post-exec binary, not the wrapper)
- The bug is silent (`2>/dev/null || true` swallows wrapper errors)
- The wrapper's auth.json corruption is immediately overwritten by kimaki's auth plugin, so the corrupted state only persists for ~milliseconds at a time — long enough to fail the next API call but short enough to be invisible to manual `cat auth.json`

Took most of a session to trace from "every restart kills OAuth" to the actual cause.

## Out of scope

- The original PR #102 (orphan-reaping at kimaki restart) is correct and helpful but addresses a *different* race (multi-worker `auth.json` writes during normal operation). This issue is about the wrapper itself being stale, not about concurrent writes.
- Whether kimaki should bypass PATH and invoke the real opencode binary directly is upstream (`remorses/kimaki`) and outside this repo's scope.

--- a/.github/workflows/shell.yml
+++ b/.github/workflows/shell.yml
@@ -31,3 +31,11 @@ jobs:
       - uses: actions/checkout@v4
       - name: Run tests/bridge-render.sh
         run: ./tests/bridge-render.sh
+
+  opencode-wrapper:
+    name: opencode wrapper regression
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Run tests/opencode-wrapper.sh
+        run: ./tests/opencode-wrapper.sh

--- a/runtimes/opencode.sh
+++ b/runtimes/opencode.sh
@@ -33,18 +33,31 @@ _install_opencode_wrapper() {
   local OPENCODE_BIN
   OPENCODE_BIN=$(command -v opencode 2>/dev/null || echo "/usr/bin/opencode")
 
-  # Don't wrap if opencode is already a wrapper script (re-runs)
-  if head -1 "$OPENCODE_BIN" 2>/dev/null | grep -q "bash"; then
-    log "OpenCode wrapper already installed — skipping"
-    return
-  fi
-
   # Find the real opencode binary (after npm install, it's in node_modules)
   local REAL_BIN
   REAL_BIN=$(readlink -f "$OPENCODE_BIN" 2>/dev/null || echo "$OPENCODE_BIN")
-  # If the resolved path is still a wrapper, dig deeper
+  # If the resolved path is still a wrapper, dig deeper. Existing installs may
+  # have old wp-coding-agents wrappers at /usr/local/bin/opencode; parse their
+  # exec target before falling back to the npm global package path.
   if head -1 "$REAL_BIN" 2>/dev/null | grep -q "bash"; then
-    REAL_BIN="/usr/lib/node_modules/opencode-ai/bin/opencode"
+    local WRAPPED_REAL_BIN
+    WRAPPED_REAL_BIN=""
+    while IFS= read -r line; do
+      case "$line" in
+        exec\ *opencode*|exec\ /*opencode*)
+          set -- $line
+          WRAPPED_REAL_BIN="${2#\'}"
+          WRAPPED_REAL_BIN="${WRAPPED_REAL_BIN%\'}"
+          WRAPPED_REAL_BIN="${WRAPPED_REAL_BIN#\"}"
+          WRAPPED_REAL_BIN="${WRAPPED_REAL_BIN%\"}"
+          ;;
+      esac
+    done < "$REAL_BIN"
+    if [ -n "$WRAPPED_REAL_BIN" ] && [ -f "$WRAPPED_REAL_BIN" ]; then
+      REAL_BIN="$WRAPPED_REAL_BIN"
+    else
+      REAL_BIN="/usr/lib/node_modules/opencode-ai/bin/opencode"
+    fi
   fi
 
   if [ ! -f "$REAL_BIN" ]; then
@@ -52,12 +65,13 @@ _install_opencode_wrapper() {
     return
   fi
 
-  log "Installing OpenCode credential sync wrapper..."
-
-  local WRAPPER_CONTENT='#!/usr/bin/env bash
+  local WRAPPER_CONTENT
+  WRAPPER_CONTENT=$(cat <<'EOF'
+#!/usr/bin/env bash
+# wp-coding-agents-opencode-wrapper-v2
 set -euo pipefail
 
-# Syncs Anthropic credentials from Kimaki'\''s account store into the format
+# Syncs Anthropic credentials from Kimaki's account store into the format
 # opencode-claude-auth reads (~/.claude/.credentials.json). Kimaki manages
 # OAuth token refresh — this wrapper forwards fresh tokens on each invocation.
 
@@ -65,7 +79,7 @@ KIMAKI_ACCOUNTS="${HOME}/.local/share/opencode/anthropic-oauth-accounts.json"
 CLAUDE_CREDENTIALS="${HOME}/.claude/.credentials.json"
 
 if [[ -f "$KIMAKI_ACCOUNTS" ]] && command -v node >/dev/null 2>&1; then
-  node -e '"'"'
+  node -e '
     const fs = require("fs");
     const path = require("path");
     try {
@@ -91,7 +105,7 @@ if [[ -f "$KIMAKI_ACCOUNTS" ]] && command -v node >/dev/null 2>&1; then
         fs.writeFileSync(claudePath, JSON.stringify(creds, null, 2), { mode: 0o600 });
       }
     } catch {}
-  '"'"' 2>/dev/null || true
+  ' 2>/dev/null || true
 fi
 
 # Legacy sync: claude creds → opencode auth.json (fallback for built-in auth)
@@ -101,15 +115,34 @@ if [[ -f "$CLAUDE_CREDENTIALS" ]] && command -v jq >/dev/null 2>&1; then
   jq "{anthropic:{type:\"oauth\",refresh:(.claudeAiOauth.refreshToken//error(\"missing\")),access:(.claudeAiOauth.accessToken//error(\"missing\")),expires:(.claudeAiOauth.expiresAt//error(\"missing\"))}}" "$CLAUDE_CREDENTIALS" > "${AUTH_DST}.tmp" 2>/dev/null && mv "${AUTH_DST}.tmp" "$AUTH_DST"
 fi
 
-exec '"${REAL_BIN}"' "$@"
-'
+exec "__REAL_BIN__" "$@"
+EOF
+)
+  WRAPPER_CONTENT=${WRAPPER_CONTENT/__REAL_BIN__/$REAL_BIN}
+
+  if [ -f "$OPENCODE_BIN" ] && [ "$(cat "$OPENCODE_BIN" 2>/dev/null)" = "$WRAPPER_CONTENT" ]; then
+    log "OpenCode wrapper already at current version — skipping"
+    return
+  fi
 
   if [ "$DRY_RUN" = true ]; then
-    echo -e "${BLUE}[dry-run]${NC} Would install OpenCode wrapper at $OPENCODE_BIN"
+    if head -1 "$OPENCODE_BIN" 2>/dev/null | grep -q "bash"; then
+      echo -e "${BLUE}[dry-run]${NC} Would replace OpenCode wrapper at $OPENCODE_BIN"
+    else
+      echo -e "${BLUE}[dry-run]${NC} Would install OpenCode wrapper at $OPENCODE_BIN"
+    fi
   else
+    if head -1 "$OPENCODE_BIN" 2>/dev/null | grep -q "bash"; then
+      local BACKUP_PATH="${OPENCODE_BIN}.bak.$(date +%Y%m%d%H%M%S)"
+      cp "$OPENCODE_BIN" "$BACKUP_PATH"
+      log "Replacing outdated OpenCode wrapper at $OPENCODE_BIN (backup: $BACKUP_PATH)"
+    else
+      log "Installing OpenCode credential sync wrapper..."
+    fi
     echo "$WRAPPER_CONTENT" > "$OPENCODE_BIN"
     chmod +x "$OPENCODE_BIN"
     log "Installed credential sync wrapper at $OPENCODE_BIN → $REAL_BIN"
+    UPDATED_ITEMS+=("opencode credential wrapper")
   fi
 }
 

--- a/tests/opencode-wrapper.sh
+++ b/tests/opencode-wrapper.sh
@@ -1,0 +1,125 @@
+#!/bin/bash
+# tests/opencode-wrapper.sh — regression tests for the Kimaki OpenCode wrapper.
+set -eu
+
+SCRIPT_DIR="$(cd "$(dirname "$0")/.." && pwd)"
+cd "$SCRIPT_DIR"
+
+PASS=0
+FAIL=0
+
+log() { :; }
+warn() { printf 'WARN: %s\n' "$*" >&2; }
+
+assert_file_contains() {
+  local label="$1" file="$2" needle="$3"
+  if grep -qF "$needle" "$file"; then
+    echo "  ok   $label"
+    PASS=$((PASS+1))
+  else
+    echo "  FAIL $label"
+    echo "       expected $file to contain: $needle"
+    FAIL=$((FAIL+1))
+  fi
+}
+
+assert_eq() {
+  local label="$1" expected="$2" actual="$3"
+  if [ "$expected" = "$actual" ]; then
+    echo "  ok   $label"
+    PASS=$((PASS+1))
+  else
+    echo "  FAIL $label"
+    echo "       expected: '$expected'"
+    echo "       actual:   '$actual'"
+    FAIL=$((FAIL+1))
+  fi
+}
+
+TMPDIR_TEST="$(mktemp -d)"
+trap 'rm -rf "$TMPDIR_TEST"' EXIT
+
+# shellcheck disable=SC1091
+source runtimes/opencode.sh
+UPDATED_ITEMS=()
+
+make_real_opencode() {
+  local real_dir="$TMPDIR_TEST/real/bin"
+  mkdir -p "$real_dir"
+  cat > "$real_dir/opencode" <<'EOF'
+#!/bin/sh
+echo real opencode "$@"
+EOF
+  chmod +x "$real_dir/opencode"
+  printf '%s/opencode' "$real_dir"
+}
+
+make_stale_wrapper() {
+  local wrapper="$1" real_bin="$2"
+  cat > "$wrapper" <<EOF
+#!/usr/bin/env bash
+set -euo pipefail
+AUTH_SRC="\${HOME}/.claude/.credentials.json"
+AUTH_DST="\${HOME}/.local/share/opencode/auth.json"
+if [[ -f "\$AUTH_SRC" ]]; then
+  mkdir -p "\$(dirname "\$AUTH_DST")"
+  cp "\$AUTH_SRC" "\$AUTH_DST"
+fi
+exec "$real_bin" "\$@"
+EOF
+  chmod +x "$wrapper"
+}
+
+echo "==> stale wrapper replacement"
+REAL_BIN="$(make_real_opencode)"
+BIN_DIR="$TMPDIR_TEST/bin"
+mkdir -p "$BIN_DIR"
+STALE_WRAPPER="$BIN_DIR/opencode"
+make_stale_wrapper "$STALE_WRAPPER" "$REAL_BIN"
+
+CHAT_BRIDGE=kimaki
+LOCAL_MODE=false
+DRY_RUN=false
+PATH="$BIN_DIR:$PATH"
+
+_install_opencode_wrapper
+
+assert_file_contains "writes current wrapper sentinel" "$STALE_WRAPPER" "# wp-coding-agents-opencode-wrapper-v2"
+assert_file_contains "preserves real binary target" "$STALE_WRAPPER" "exec \"$REAL_BIN\""
+assert_eq "backs up stale wrapper" "1" "$(ls "$BIN_DIR"/opencode.bak.* 2>/dev/null | wc -l | tr -d ' ')"
+
+_install_opencode_wrapper
+
+assert_eq "current wrapper rerun is idempotent" "1" "$(ls "$BIN_DIR"/opencode.bak.* 2>/dev/null | wc -l | tr -d ' ')"
+
+echo "==> skip gates"
+NON_KIMAKI_DIR="$TMPDIR_TEST/non-kimaki/bin"
+mkdir -p "$NON_KIMAKI_DIR"
+NON_KIMAKI_WRAPPER="$NON_KIMAKI_DIR/opencode"
+make_stale_wrapper "$NON_KIMAKI_WRAPPER" "$REAL_BIN"
+CHAT_BRIDGE=telegram
+LOCAL_MODE=false
+PATH="$NON_KIMAKI_DIR:$PATH"
+_install_opencode_wrapper
+assert_eq "non-kimaki bridge leaves wrapper alone" "0" "$(grep -c 'wp-coding-agents-opencode-wrapper-v2' "$NON_KIMAKI_WRAPPER")"
+
+LOCAL_DIR="$TMPDIR_TEST/local/bin"
+mkdir -p "$LOCAL_DIR"
+LOCAL_WRAPPER="$LOCAL_DIR/opencode"
+make_stale_wrapper "$LOCAL_WRAPPER" "$REAL_BIN"
+CHAT_BRIDGE=kimaki
+LOCAL_MODE=true
+PATH="$LOCAL_DIR:$PATH"
+_install_opencode_wrapper
+assert_eq "local mode leaves wrapper alone" "0" "$(grep -c 'wp-coding-agents-opencode-wrapper-v2' "$LOCAL_WRAPPER")"
+
+echo "==> upgrade phase guard"
+assert_file_contains "upgrade sources opencode runtime for kimaki" upgrade.sh 'source "$SCRIPT_DIR/runtimes/opencode.sh"'
+assert_file_contains "upgrade invokes wrapper refresh" upgrade.sh "_install_opencode_wrapper"
+
+echo
+if [ "$FAIL" -gt 0 ]; then
+  echo "FAILED: $FAIL of $((PASS+FAIL)) assertion(s)"
+  exit 1
+fi
+echo "OK: $PASS / $PASS assertions passed"

--- a/upgrade.sh
+++ b/upgrade.sh
@@ -626,12 +626,30 @@ update_chat_bridge_launchd() {
 reapply_claude_auth_patch() {
   _run_filter_active patch || return 0
 
-  if [ "$RUNTIME" != "opencode" ]; then
-    log "Phase 7: Skipping (runtime is $RUNTIME, not opencode)"
+  if [ "$RUNTIME" != "opencode" ] && [ "$CHAT_BRIDGE" != "kimaki" ]; then
+    log "Phase 7: Skipping (runtime is $RUNTIME and chat bridge is $CHAT_BRIDGE)"
     return 0
   fi
 
-  log "Phase 7: Re-applying opencode-claude-auth PascalCase patch..."
+  log "Phase 7: Checking OpenCode auth integration..."
+
+  if ! declare -F _install_opencode_wrapper >/dev/null; then
+    # Kimaki spawns opencode-serve even when the primary editing runtime is not
+    # OpenCode, so stale wrappers must be repairable from any Kimaki install.
+    # Source the runtime file for its wrapper helper without running a full
+    # OpenCode runtime install.
+    # shellcheck disable=SC1091
+    source "$SCRIPT_DIR/runtimes/opencode.sh"
+  fi
+
+  _install_opencode_wrapper
+
+  if [ "$RUNTIME" != "opencode" ]; then
+    log "  Skipping opencode-claude-auth patch (runtime is $RUNTIME)"
+    return 0
+  fi
+
+  log "  Re-applying opencode-claude-auth PascalCase patch..."
 
   if [ ! -f "$SCRIPT_DIR/lib/patch-claude-auth.py" ]; then
     warn "  patch-claude-auth.py not found — skipping"

